### ISSUE TITLE
"configuration not public" on local test->test dependency if the same version has been resolved out of a public Maven repository

### ISSUE THREAD

**Issue by mosesn** [sbt/sbt#1827]
# Problem

I saw this problem on [stackoverflow](http://stackoverflow.com/questions/26235997/teamcity-cant-resolve-test-artifact-from-nexus-with-configuration-not-public) too, but it didn't seem like it had been solved, and I couldn't find an issue, so I'm posting it here.

I'm trying to do development on twitter/util, but when I try to run update on util-collection, it tells me: 

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.twitter#util-core_2.10;6.23.0: configuration not public in com.twitter#util-core_2.10;6.23.0: 'test'. It was required from com.twitter#util-collection_2.10;6.23.0 test
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]
[warn]  Note: Unresolved dependencies path:
[warn]      com.twitter:util-core_2.10:6.23.0
[warn]        +- com.twitter:util-collection_2.10:6.23.0
[trace] Stack trace suppressed: run last util-collection/*:update for the full output.
[error] (util-collection/*:update) sbt.ResolveException: unresolved dependency: com.twitter#util-core_2.10;6.23.0: configuration not public in com.twitter#util-core_2.10;6.23.0: 'test'. It was required from com.twitter#util-collection_2.10;6.23.0 test
[error] Total time: 1 s, completed Jan 27, 2015 1:49:54 AM
```

I tried making a demo project to repro, and I found that I could only repro when I chose the same version, organization, and artifact name.

**Comment by mosesn** [comment]
More interesting stuff:

```
 mnakamura ⮀ ~ ⮀ find ~/.ivy2/ -name 6.23.0
/Users/mnakamura/.ivy2//local/com.twitter/util-core_2.10/6.23.0
```

**Comment by eed3si9n** [comment]
What's the repro steps using twitter util 6.23.0?
I grabbed the code, ran `publishLocal`.

```
$ find ~/.ivy2/ -name 6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-app_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-benchmark_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-cache_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-class-preloader_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-codec_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-collection_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-core_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-events_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-hashing_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-jvm_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-logging_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-reflect_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-stats_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-test_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-thrift_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-zk-common_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util-zk_2.10/6.23.0
/Users/eugene/.ivy2//local/com.twitter/util_2.10/6.23.0
```

I started sbt again, switch into util-collection and run all the tests:

```
[info] GenerationalQueueTest:
[info] - ExactGenerationalQueue: Don't collect fresh data
[info] - ExactGenerationalQueue: Don't collect old data recently refreshed
[info] - ExactGenerationalQueue: collect one old data
[info] - ExactGenerationalQueue: collectAll old data
[info] - ExactGenerationalQueue: generate a long chain of bucket (if applicable)
[info] - ExactGenerationalQueue: Don't collect old data
[info] - BucketGenerationalQueue: Don't collect fresh data
[info] - BucketGenerationalQueue: Don't collect old data recently refreshed
[info] - BucketGenerationalQueue: collect one old data
[info] - BucketGenerationalQueue: collectAll old data
[info] - BucketGenerationalQueue: generate a long chain of bucket (if applicable)
[info] - BucketGenerationalQueue: Don't collect old data
[info] - BucketGenerationalQueue check removals
[info] ScalaTest
[info] Run completed in 302 milliseconds.
[info] Total number of tests run: 13
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 13, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Passed: Total 13, Failed 0, Errors 0, Passed 13
```

I would in general recommend against publishing final version number locally since things will get cached by Ivy, but as-is I'm not able to reproduce the problem.

**Comment by mosesn** [comment]
@eed3si9n could you try purging your local published version and just downloading the regular version off of maven central?

**Comment by eed3si9n** [comment]
I used `cleanLocal` to wipe out the util:

```
$ find ~/.ivy2/ -name 6.23.0
$ 
```

You said:

> when I try to run update on util-collection

so I ran `clean` and then `update` from `util-collection`. Works fine.
util-core resolves as an internal dependency, so it doesn't show up as an external dependency for me:

```
util-collection> update
[info] Updating {file:/Users/eugene/work/quick-test/twitter-util/}util-collection...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[success] Total time: 1 s, completed Feb 24, 2015 12:04:26 AM
util-collection> show externalDependencyClasspath
[info] List(Attributed(doesnotexist), Attributed(/Users/eugene/.conscript/boot/scala-2.10.4/lib/scala-library.jar), Attributed(/Users/eugene/.ivy2/cache/com.google.code.findbugs/jsr305/jars/jsr305-1.3.9.jar), Attributed(/Users/eugene/.ivy2/cache/javax.inject/javax.inject/jars/javax.inject-1.jar), Attributed(/Users/eugene/.ivy2/cache/com.google.guava/guava/bundles/guava-16.0.1.jar), Attributed(/Users/eugene/.ivy2/cache/commons-collections/commons-collections/jars/commons-collections-3.2.1.jar))
[success] Total time: 0 s, completed Feb 24, 2015 12:04:41 AM
```

**Comment by mosesn** [comment]
Hmm, ok, could you cleanLocal in util again, clone twitter/finagle, run update, make sure that util-core is downloaded, and then run util-collection update?

**Comment by eed3si9n** [comment]
Grabbing finagle did it.

```
util-collection> update
[info] Updating {file:/Users/eugene/work/quick-test/twitter-util/}util-core...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Updating {file:/Users/eugene/work/quick-test/twitter-util/}util-collection...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.twitter#util-core_2.10;6.23.0: configuration not public in com.twitter#util-core_2.10;6.23.0: 'test'. It was required from com.twitter#util-collection_2.10;6.23.0 test
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      com.twitter:util-core_2.10:6.23.0
[warn]        +- com.twitter:util-collection_2.10:6.23.0
[trace] Stack trace suppressed: run last util-collection/*:update for the full output.
[error] (util-collection/*:update) sbt.ResolveException: unresolved dependency: com.twitter#util-core_2.10;6.23.0: configuration not public in com.twitter#util-core_2.10;6.23.0: 'test'. It was required from com.twitter#util-collection_2.10;6.23.0 test
[error] Total time: 1 s, completed Feb 24, 2015 12:42:55 AM
```

In my Ivy cache now is ivy-6.23.0.xml, which I guess was translated from pom saying the test configuration is private:

```
<conf name="test" visibility="private" description="this scope indicates that the dependency is not required for normal use of the application, and is only available for the test compilation and execution phases." extends="runtime"/>
```

**Comment by eed3si9n** [comment]
I guess a workaround would be using git SHA versioning or SNAPSHOT for non finals.

**Comment by mosesn** [comment]
We have a "develop" branch and a "master" branch, where "develop" can depend on unpublished code in other twitter libraries (and hence requires locally publishing other libs) and, and where "master" only depends on stable builds.  We could bump the version number, but since master is typically the last published stable build, it's nice to be able to use the actual version number so you can see the code that's on the last version, and it's clear that it represents the last version.

We use -SNAPSHOT on the "develop" branch, which means that only folks who aren't on the "develop" branch run into the problem, but it would be nice to be able to build "master" too, without bumping the version or adding -SNAPSHOT suffix.  With that said, you're right that if it becomes a real pain point that that's a good workaround.

Thanks for taking the time to reproduce, sorry my reproduction case was so inexact.

**Comment by eed3si9n** [comment]
Ideally speaking when you're working on `util-collection`, it _shouldn't_ matter what's in ivy cache. Since it depends on `util-core` locally we should prioritize that first. In that sense this is a sbt bug. Not sure how easy it would be to fix this, but I'll keep this open.

**Comment by mosesn** [comment]
:+1: awesome, thanks!  I'll keep following it, let me know what I can do to help.

**Comment by jsuereth** [comment]
Wow, yeah, this is where the impedance mismatch between Ivy + maven rears its ugly head.  What happens here is we've LOST information going Ivy -> Maven and back.  I'll try to dive into the issue shortly, but what we need is a new implementation of the Ivy cache that can evict "bad" nodes like this...

**Comment by greenrd** [comment]
I am experiencing this issue with -SNAPSHOT dependencies.

**Comment by eed3si9n** [comment]
I'm running into this now as I'm trying to do `"compile->compile;test->test"` using a Maven style repository. It seems like when Maven's `"test"` scope is translated into an Ivy configuration, Ivy makes it a private one.

https://github.com/sbt/ivy/blob/3cf31483b03e0ba294c76f5391dc7570b7649090/src/java/org/apache/ivy/plugins/parser/m2/PomModuleDescriptorBuilder.java#L92-L96:

``` java
        new Configuration("test", Visibility.PRIVATE,
                "this scope indicates that the dependency is not required for normal use of "
                + "the application, and is only available for the test compilation and "
                + "execution phases.",
                new String[] {"runtime"}, true, null),
```

Modifying Ivy to make that public may have other consequences like `update` trying to resolve normally-unnecessary test dependencies. Not sure.

Here's my current workaround:

``` scala
    libraryDependencies ++= Seq(utilLogging, (utilLogging % Test).classifier("tests")),
```

**Comment by mikepii** [comment]
For others who may come to this page from Google: sometimes clearing out ~/.ivy2/cache/ can fix this when trying to build twitter-util ([someone posted this in google groups](https://groups.google.com/forum/#!topic/finaglers/9R6UNgWe2N8))

**Comment by dragos** [comment]
This is pretty easy to run into when building Spark (fixed by `rm -rf ~/.ivy2/cache`):

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: org.apache.spark#spark-network-common_2.11;1.5.1: configuration not public in org.apache.spark#spark-network-common_2.11;1.5.1: 'test'. It was required from org.apache.spark#spark-network-shuffle_2.11;1.5.1 test
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      org.apache.spark:spark-network-common_2.11:1.5.1 ((com.typesafe.sbt.pom.MavenHelper) MavenHelper.scala#L76)
[warn]        +- org.apache.spark:spark-network-shuffle_2.11:1.5.1
[trace] Stack trace suppressed: run last network-shuffle/*:update for the full output.
[error] (network-shuffle/*:update) sbt.ResolveException: unresolved dependency: org.apache.spark#spark-network-common_2.11;1.5.1: configuration not public in org.apache.spark#spark-network-common_2.11;1.5.1: 'test'. It was required from org.apache.spark#spark-network-shuffle_2.11;1.5.1 test
```

**Comment by eed3si9n** [comment]
@jsuereth How should we resolve this one?  Some ideas that's been mentioned thus far are
- "a new implementation of the Ivy cache"
- Prioritize inter-project resolution over Ivy cache.

Here's another idea:
- Display a warning saying `test->test` won't work if `publishMavenStyle` is `true`

Isn't allowing `test->test` promising something we can't deliver since Test configuration is private? (See also `intransitive()`) What do you think?

**Comment by maheshkelkar** [comment]
Sorry, does anybody have a `hacky` workaround for this?

**Comment by eed3si9n** [comment]
@maheshkelkar Are you seeing this with sbt 0.13.11?

**Comment by maheshkelkar** [comment]
@eed3si9n: Yes. I was using 0.13.8, but bumped it up to 0.13.11, but same problem `http://github.com/lookout/borderpatrol`. The SNAPSHOT JARs are available on jfrog. BTW if I use locally published version of this repo, it works fine.

In my internal repository, I am trying to use the `test` repositories as follows:

```
val testDependencies = Seq(
  "com.lookout" %% "borderpatrol-core" % borderPatrolVersion % "test->test",
  "com.lookout" %% "borderpatrol-auth" % borderPatrolVersion % "test->test"
)
lazy val test = project
  .settings(moduleName := "borderpatrol-service-test")
  .settings(allSettings)
  .settings(noPublish)
  .settings(libraryDependencies ++= testDependencies)
```

Here is the error:

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.lookout#borderpatrol-core_2.11;0.2.4-SNAPSHOT: configuration not public in com.lookout#borderpatrol-core_2.11;0.2.4-SNAPSHOT: 'test'. It was required from com.lookout#borderpatrol-service-test_2.11;0.1.14-SNAPSHOT test
[warn]  :: com.lookout#borderpatrol-auth_2.11;0.2.4-SNAPSHOT: configuration not public in com.lookout#borderpatrol-auth_2.11;0.2.4-SNAPSHOT: 'test'. It was required from com.lookout#borderpatrol-service-test_2.11;0.1.14-SNAPSHOT test
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::

```

**Comment by mosesn** [comment]
Someone recently ran into this issue again and mentioned it on the finagle gitter.  Is there any chance this will get scheduled?

**Comment by eed3si9n** [comment]
@mosesn Is that before or after https://github.com/twitter/finagle/commit/6c8f8e23e30e83789a6561c7cb949588107a865c? https://github.com/sbt/sbt/pull/2345 was released in 0.13.11 (Feb 2016).

**Comment by mosesn** [comment]
Before.  I'll check in with the person with the problem to see if this fixes it, thanks!

**Comment by eed3si9n** [comment]
Closing this since this should be addressed by #2345 as much as we can.
Ultimately, `test->test` is not really safe for any artifacts published to Maven.
